Kh/fix/pwa for android

bromiesTM opened the pull request.

bromiesTM committed:
c/feat: add minimal service worker for PWA installation support

- Add basic service worker with install/activate events
- Minimal implementation for Android PWA prompt requirements
- No fetch handling to keep service worker lightweight

bromiesTM committed:
c/feat: enhance PWA detection and platform support

- Improve Safari detection to exclude Chrome on iOS
- Better PWA installation criteria checking
- Enhanced Android and iOS compatibility for PWA prompts

bromiesTM committed:
c/feat: integrate service worker registration in app layout

- Register service worker on app startup for PWA functionality
- Set up PWA event listeners early in app lifecycle
- Clean implementation without debug console logs

bromiesTM committed:
c/fix: improve PWA notification timing

thlehmann-ionos commented on src/lib/IONOS/stores/pwa-prompt.ts:
```diff
@@ -24,14 +36,25 @@ export function setupGlobalPWAListener(): (() => void) | null {
-	window.onbeforeinstallprompt = handleBeforeInstallPrompt;
+
+	if ('onbeforeinstallprompt' in window) {
+		window.onbeforeinstallprompt = handleBeforeInstallPrompt;
```
Why do you define it in addition to `addEventListener()`?

> Use the event name in methods like [addEventListener()](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener), **or set an event handler property.**

https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeinstallprompt_event#syntax

bromiesTM replied:
probably leftover from testing, i will remove it

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -71,29 +71,38 @@
 
 	onMount(() => {
-		cleanupPWAListeners = setupGlobalPWAListener();
+		// Add delay to allow service worker registration to complete
```
Can we be sure 1s is always sufficient?

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -71,29 +71,38 @@
-		notifications.update((currentNotifications: Notification[]) =>
-			currentNotifications.filter(n => n.type !== NotificationType.PWA_INSTALL)
-		);
+	$: {
```
This seems to be pure syntactical change. Please undo if not needed or commit separately as "chore:" or "style:".

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -71,29 +71,38 @@
 	}
 
 	const addPWANotification = () => {
+		// Check if PWA notification already exists to avoid duplicates
```
Can we instead avoid adding it twice?

thlehmann-ionos commented on src/lib/IONOS/components/notifications/NotificationManager.svelte:
```diff
@@ -71,29 +71,38 @@
 
-		if (shouldShowPWAPrompt($deferredPrompt)) {
+	$: {
+		if ($isPWAInstallable && shouldShowPWAPrompt($deferredPrompt)) {
```
Why is it added in two places, here and in `onMount()`?

thlehmann-ionos commented on src/lib/IONOS/stores/pwa-prompt.ts:
```diff
@@ -6,6 +6,18 @@ export const isPWAInstallable = writable(false);
+let earlyPromptEvent: BeforeInstallPromptEvent | null = null;
+let earlyListenerAdded = false;
+
+if (typeof window !== 'undefined' && !earlyListenerAdded) {
```
1. Modules should not execute code as it makes the execution order dependent on where the module is included first and it's harder or no longer testable
2. Why is this done in addition to `setupGlobalPWAListener()`

thlehmann-ionos reviewed (commented):
Let's discuss the review comments.

Also: according to [this documentation][1] the `manifest.json` should contain more details

[1]: https://web.dev/articles/install-criteria

bromiesTM committed:
chore(pwa): remove duplicate event handling

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

thlehmann-ionos reviewed (changes requested):
Let's discuss.

# The banner is randomly sometimes not shown 

* 50:50
* Sometimes shown multiple times in a row, sometimes not multiple times in a row 
* I see no obvious race condition
* This happened with the previous revision (cec865c) too (so it's not a regression due to your last change)

I suspect it's just too late to listen for this event, even in `+layout.svelte` (the module itself too, because it will be loaded by some other component, which might happen too late).


# The text is not translated 

This is due to i18n being initialized in root `+layout.svelte` _after_ that translation was requested.

Queuing the message should be deferred until i18n is ready.


# Not functional in Dev-Environment

As you pointed out it's not locally testable with the dev-server. This is not just some nice to have. 20mins - 25mins iterations cycles to test a change is unacceptable.

I checked a couple of possible reasons, yet I see no obvious ones. I suspect the much slower load times may hinder proper execution order (which would relate to issue no. 1) or some headers set by the Vite server may prevent some PWA features.

The app would be installable as hinted by the Install button in the URL bar (or "add to home screen dialog"). This further points to my 1st suspicion.

bromiesTM committed:
c/fix(pwa): fix pwa & service worker race condition

Use SvelteKit Service Worker that registers automatically
Use SvelteKit Client-Hook to make sure the beforeinstall event is listened to as early as possbile

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
c/fix(notifications): remove notifications by id

`removeNotification()` was broken as it cant compare object identity

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

bromiesTM committed:
c/fix(pwa): notification text appears in correct language

Adding the notification too early ment the i18n translation was not yet loaded properly.
The notification object is now a subscribable store to make sure the translations are working at render time

Signed-off-by: Kai Henseler <kai.henseler@strato.de>

thlehmann-ionos reviewed (approved):
## `c/fix(pwa): fix pwa & service worker race condition`

* Please remove the `console.log`, otherwise LTGM


## `c/fix(notifications): remove notifications by id`

* Please mention in the commit message that the filter in `removeNotification()` was broken as it can not compare object identity, otherwise LTGM

thlehmann-ionos commented:
Please mind #97.
